rpc/query_server.go

```diff
@@ -1094,8 +1094,38 @@ func (s *QueryServer) EthGetStorageAt(local eth.Data, position string, block eth
 	return eth.EncBytes(storage), nil
 }
 
-func (s *QueryServer) EthEstimateGas(query eth.JsonTxCallObject) (eth.Quantity, error) {
-	return eth.Quantity("0x0"), nil
+// https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_estimateGas
+func (s *QueryServer) EthEstimateGas(query eth.JsonTxCallObject, block eth.BlockHeight) (resp eth.Quantity, err error) {
+	var caller loom.Address
+	const ethChainID = "eth"
+	if len(query.From) > 0 {
+		caller, err = eth.DecDataToAddress(ethChainID, query.From)
+		if err != nil {
+			return resp, err
+		}
+	} else {
+		caller = loom.RootAddress(ethChainID)
+	}
+
+	contract, err := eth.DecDataToAddress(s.ChainID, query.To)
+	if err != nil {
+		return resp, err
+	}
+
+	data, err := eth.DecDataToBytes(query.Data)
+	if err != nil {
+		return resp, err
+	}
+
+	snapshot := s.StateProvider.ReadOnlyState()
+	defer snapshot.Release()
+
+	vm := levm.NewLoomVm(snapshot, nil, nil, nil, true)
```

Comment:
Do we really need debug mode enabled in the EVM to estimate gas?

Reply:
Also, if we don't pass the account balance manager to `NewLoomVm` then any call to a payable function, or a function that does a balance check is likely to produce an inaccurate gas estimate since the code path that will be executed is likely to be quite different from the one that will be executed when ETH is available.